api/python/quilt3/data_transfer.py

```diff
@@ -991,81 +918,114 @@ def _calculate_sha256_internal(src_list, sizes, results):
+            if result is not None and not isinstance(result, Exception):
+                futures.append((False, []))
+            elif size < s3_transfer_config.multipart_threshold:
+                future = executor.submit(_process_url_part, src, 0, size)
+                futures.append((False, [future]))
+            else:
+                adjuster = ChunksizeAdjuster()
+                chunksize = adjuster.adjust_chunksize(s3_transfer_config.multipart_chunksize, size)
+
+                src_future_list = []
+                for start in range(0, size, chunksize):
+                    end = min(start + chunksize, size)
+                    future = executor.submit(_process_url_part, src, start, end-start)
+                    src_future_list.append(future)
+
+                futures.append((True, src_future_list))
 
-    with tqdm(desc="Hashing", total=total_size, unit='B', unit_scale=True, disable=DISABLE_TQDM) as progress, \
-         ThreadPoolExecutor() as executor, \
-         ThreadPoolExecutor(
-             MAX_CONCURRENCY,
-             thread_name_prefix='s3-executor',
-         ) as s3_executor:
-        s3_context = types.SimpleNamespace(
-            find_correct_client=with_lock(S3ClientProvider().find_correct_client),
-            pending_parts_semaphore=threading.BoundedSemaphore(s3_max_pending_parts),
-            executor=s3_executor,
-        )
-        progress_update = with_lock(progress.update)
-        future_to_idx = {
-            executor.submit(_process_url, src, size): i
-            for i, (src, size, result) in enumerate(zip(src_list, sizes, results))
-            if result is None or isinstance(result, Exception)
-        }
         try:
-            for future in concurrent.futures.as_completed(future_to_idx):
-                results[future_to_idx.pop(future)] = future.result()
+            for idx, (is_multipart, future_list) in enumerate(futures):
+                future_results = [future.result() for future in future_list]
+                exceptions = [ex for ex in future_results if isinstance(ex, Exception)]
+                if exceptions:
+                    results[idx] = exceptions[0]
+                elif not future_results:
+                    assert results[idx] is not None and not isinstance(results[idx], Exception)
+                elif not is_multipart:
+                    results[idx] = base64.b64encode(future_results[0]).decode()
+                else:
+                    hashes_hash = hashlib.sha256(b''.join(future_results)).digest()
+                    results[idx] = f'{base64.b64encode(hashes_hash).decode()}-{len(future_results)}'
         finally:
             stopped = True
-            while future_to_idx:
-                future, idx = future_to_idx.popitem()
-                future.cancel()
+            for _, future_list in futures:
+                for future in future_list:
+                    future.cancel()
 
     return results
 
 
+def calculate_sha256_bytes(data: bytes):
```

Comment:
It's weird we have `calculate_sha256()` and `calculate_sha256_bytes()` and they calculate something different.